presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java

```diff
@@ -78,6 +97,84 @@ public List<String> getEffectivePartitions(
                 partitionPredicate.put(partitionColumn, Domain.all(column.getHiveType().getType(typeManager)));
             }
         }
-        return metastore.getPartitionNamesByFilter(metastoreContext, schemaName, tableName, partitionPredicate);
+        // fetch partition names from metastore
+        List<String> partitionNames = metastore.getPartitionNamesByFilter(metastoreContext, schemaName, tableName, partitionPredicate);
+        List<Type> partitionTypes = partitionColumns.stream()
+                .map(column -> column.getType().getType(typeManager))
+                .collect(toList());
+        // match partition and get partition only we need
+        Iterable<HivePartition> partitionsIterable = () -> partitionNames.stream()
+                // Apply extra filters which could not be done by getFilteredPartitionNames
+                .map(partitionName -> parseValuesAndFilterPartition(new SchemaTableName(schemaName, tableName), partitionName, hudiColumnHandles, partitionTypes, new Constraint<>(tupleDomain)))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .iterator();
+        return getPartitionsAsList(partitionsIterable.iterator()).stream()
+                .map(HivePartition::getPartitionId)
+                .collect(toList());
+    }
+
+    private Optional<HivePartition> parseValuesAndFilterPartition(
+            SchemaTableName tableName,
+            String partitionId,
+            List<HudiColumnHandle> partitionColumns,
+            List<Type> partitionColumnTypes,
+            Constraint<ColumnHandle> constraint)
+    {
+        HivePartition partition = parsePartition(tableName, partitionId, partitionColumns, partitionColumnTypes, timeZone);
+
+        Map<ColumnHandle, Domain> domains = constraint.getSummary().getDomains().get();
+        for (HudiColumnHandle column : partitionColumns) {
+            NullableValue value = partition.getKeys().get(column);
+            Domain allowedDomain = domains.get(column);
+            if (allowedDomain != null && !allowedDomain.includesNullableValue(value.getValue())) {
+                return Optional.empty();
+            }
+        }
+
+        if (constraint.predicate().isPresent() && !constraint.predicate().get().test(partition.getKeys())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(partition);
+    }
+
+    public static HivePartition parsePartition(
```

Comment:
Can we use `HudiPartition` we already have instead of `HivePartition`? Using two partition kinds might make it confusing to other developers.

Reply:
You are right. I have used HudiPartition instead of HivePartition.